packages/dd-trace/src/native/metrics/main.cc

```diff
@@ -0,0 +1,468 @@
+  }
+
+  static void OnCheck(uv_check_t* handle) {
+    auto self = reinterpret_cast<EventLoopStat*>(handle->data);
+
+    uint64_t check_time = uv_hrtime();
+    uint64_t poll_time = check_time - self->prepare_time_;
+    uint64_t latency = self->prepare_time_ - self->check_time_;
+    uint64_t timeout = self->timeout_ * 1000 * 1000;
+
+    if (poll_time > timeout) {
+      latency += poll_time - timeout;
+    }
+
+    self->histogram_.Add(latency);
+    self->check_time_ = check_time;
+  }
+
+  void Start() {
+    uv_prepare_start(&prepare_handle_, OnPrepare);
+    uv_check_start(&check_handle_, OnCheck);
+  }
+
+  void Stop() {
+    uv_prepare_stop(&prepare_handle_);
+    uv_check_stop(&check_handle_);
+    histogram_.Reset();
+  }
+
+  void Dump(Env env, Object& o) {
+    o["eventLoop"] = histogram_.ToValue(env);
+    histogram_.Reset();
+  }
+
+  uv_check_t check_handle_;
+  uv_prepare_t prepare_handle_;
+  uint64_t check_time_;
+  uint64_t prepare_time_;
+  uint64_t timeout_;
+  Histogram histogram_;
+};
+
+
+  /////////////////////
+ // HEAP STATISTICS //
+/////////////////////
+class HeapStat : public Collector {
+ public:
+  ~HeapStat() { Stop(); }
+
+ private:
+  void Start() {}
+  void Stop() {}
+
+  void Dump(Env env, Object& o) {
+    v8::Isolate* isolate = v8::Isolate::GetCurrent();
+    v8::HeapSpaceStatistics stats;
+
+    auto heap = Object::New(env);
+    auto spaces = Array::New(env);
+    heap["spaces"] = spaces;
+
+    for (size_t i = 0; i < isolate->NumberOfHeapSpaces(); i += 1) {
+      v8::HeapSpaceStatistics stats;
+      auto space = Object::New(env);
+      if (isolate->GetHeapSpaceStatistics(&stats, i)) {
+        space["space_name"] = std::string(stats.space_name());
+        space["space_size"] = stats.space_size();
+        space["space_used_size"] = stats.space_used_size();
+        space["space_available_size"] = stats.space_available_size();
+        space["physical_space_size"] = stats.physical_space_size();
+
+        spaces[i] = space;
+      } else {
+        spaces[i] = env.Null();
+      }
+    }
+
+    o["heap"] = heap;
+  }
+};
+
+static uint64_t time_to_micro(uv_timeval_t timeval) {
+  return timeval.tv_sec * 1000 * 1000 + timeval.tv_usec;
+}
+
+class ProcessStat : public Collector {
+ public:
+  ~ProcessStat() { Stop(); }
+
+ private:
+  void Start() {}
+  void Stop() {}
+
+  void Dump(Env env, Object& o) {
+    uv_rusage_t usage;
+    uv_getrusage(&usage);
+
+    auto cpu = Object::New(env);
+    cpu["user"] = time_to_micro(usage.ru_utime) - time_to_micro(usage_.ru_utime);
+    cpu["system"] = time_to_micro(usage.ru_stime) - time_to_micro(usage_.ru_stime);
+
+    o["cpu"] = cpu;
+
+    usage_ = usage;
+  }
+
+  uv_rusage_t usage_;
+};
+
+
+  ///////////////////
+ // SPAN TRACKING //
+///////////////////
+class SpanTracker : public Collector {
+ public:
+  ~SpanTracker() {
+    Stop();
+  }
+
+  uint64_t Track(const Object& span) {
+    if (!running_) {
+      return 0;
```

Comment:
Since `id_counter_` starts at `0`, that could cause issues when tracking is started/stopped since `0` has a special meaning. It would be safer to use an out of range value like -1.